eclipse-theia/theia
# GH-6428: Refactored the built-in monaco commands.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>



#### What it does
Changes the way how we handle monaco commands, and hopefully providers a better UX for the `Undo`, `Redo`, and `Select All` commands. It is the fix for electron, as it 

#### How to test

### Test Plan:

- Preconditions:
   - If you are verify this in electron, you have to be avere of the non-dynamic nature of the application menu. (https://github.com/eclipse-theia/theia/issues/6425).
   - Run `Reset Workbench Layout`.
   - Open the SIW widget; the `input` widget has the focus.
- Steps:
   - Type _**foo**_ and press <kbd>Ctrl/⌘</kbd>+<kbd>A</kbd> (`Select All`). Assert: the `input` value is _**foo**_ and is selected.
   - Press <kbd>Del</kbd>. Assert: the `input` is empty.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>Z</kbd> (`Undo`). Assert: the `input` value is _**foo**_ and is selected.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd> (`Redo`). Assert: the `input` is empty.
   - Type _**bar**_ and select the `Selection` > `Select All` menu item. Assert: the input value is _**bar**_ and is selected.
   - Press <kbd>Del</kbd>. Assert: the `input` is empty.
   - Select the `Edit` > `Undo` menu item. Assert: the input is _**bar**_ and is selected. (Note: there is a bug here. The SIW widget's `input` does not listen on value change and does not re-run the query.)
   - Select the `Edit` > `Redo` menu item. Assert: the input is empty.

-----

- Preconditions:
   - If you are verify this in electron, you have to be avere of the non-dynamic nature of the application menu. (https://github.com/eclipse-theia/theia/issues/6425).
   - Open `theia` as a workspace in Theia.
   - Run `Reset Workbench Layout`.
   - Open the SIW widget; the `input` widget has the focus, type _**baz**_.
   - Open the `os.ts` module via the _Command Palette_. If it is opened in a preview mode, make sure the editor has the focus; the cursor blinks inside the editor.
- Steps:
   - Type _**foo**_ somewhere in the editor and press <kbd>Ctrl/⌘</kbd>+<kbd>Z</kbd> (`Undo`). Assert: _**foo**_ is not there.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>Z</kbd> (`Redo`). Assert: _**foo**_ is there.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>Z</kbd> (`Undo`). Assert: _**foo**_ is not there.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>A</kbd> (`Select All`). Assert: the editor content is selected.
   - Press <kbd>Esc</kbd>. Assert: the cursor blinks at the end of the file, nothing is selected.
   - Type _**bar**_ and select the `Edit` > `Undo` menu item. Assert: _**bar**_ is not there.
   - Select the `Edit` > `Redo` menu item. Assert: _**bar**_ is in the editor.
   - Set the focus (single-click) into the SIW's `input`.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>A</kbd> (`Select All`). Assert: _**baz**_ is selected.
   - Make sure the cursor blinks once again in the editor and select the `Selection` > `Select All` menu item. Assert: the editor content is selected.
   - Press <kbd>Esc</kbd>. Assert: the cursor blinks at the end of the file, nothing is selected.
   - Right-click on `isWindows` in the editor and select `Peek` > `Peek References` from the context menu. Assert: embedded editor is open. The cursor blinks on `isWindows` in the main editor.
   - Press <kbd>Ctrl/⌘</kbd>+<kbd>A</kbd> (`Select All`). Assert: the main editor content is selected.
   - Make sure the cursor blinks this time in the embedded editor and press <kbd>Ctrl/⌘</kbd>+<kbd>A</kbd> (`Select All`). Assert: the embbedded editor content is selected. (Fixes #7483)

----

🎖Let's verify other issues:
 - Inside the `os.ts` module, press <kbd>Ctrl/⌘</kbd>+<kbd>F</kbd>, type `is` into the search widget. If you press <kbd>Enter</kbd> (multiple times) the selection jumps to the next `is` occurrence position in the editor. Bonus: it works inside the embedded editor too. (Fixes #7460) 
 - If you are verifying the changes with the electron example (Fixes #6428), open the _Command Palette_ and make sure you can do `Undo`, `Redo`, and `Select All` from the keyboard. (Fixes #2756)
 - Close all editors and press <kbd>Ctrl/⌘</kbd>+<kbd>F</kbd>. Assert: no errors because of the missing handler for `core.find` command. (Fixes #7474)


----
TODOs:
 - [x] Verify [this](https://github.com/microsoft/vscode/issues/88610). From time to time, I can see the `Commiting pending changes before change accessor leaves due to read access.` warning when running the electron example. Hmm, but the logger code was removed by the VS Code guys in January. Let's double-check it. Verified ✅ (https://github.com/microsoft/vscode/commit/6beb75765e492b47e32554e093b77317fec397a8)

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)


@akosyakov, @RomanNikitenko, here is an early preview of my changes. I am going to verify the behavior locally and write down the verification steps. Please take a look at my changes. Thank you!
